File: benchmark/harness/Sources/BenchmarkKit/Report.swift
// Report.swift — render benchmark/results/history.json into a self-contained
// result.html. Port of benchmark/lib/report.py.
//
// Shows ALL metrics (WITH vs WITHOUT + deltas) for every retained run, per mode,
// newest first, plus each generated app's path, derived analysis rows (each
// degrades to an em-dash when inputs are missing), coverage rollup rows for live
// records that carry a per-stage coverage manifest, and a plugin-surface coverage
// ratio (exercised/declared) computed at report time from the repo's
// agents/*.md, commands/*.md, and skills/*/SKILL.md counts.

import Foundation

public enum Report {
    // (accessor, label, isDeltaNumeric)
    static let metricRows: [(String, String, Bool)] = [
        ("tokens.in", "tokens in", true),
        ("tokens.out", "tokens out", true),
        ("tokens.total", "tokens total", true),
        ("cost_usd", "cost (USD)", true),
        ("wall_clock_s", "wall clock (s)", true),
        ("loc_produced", "LOC produced", true),
        ("test_count", "test count", true),
        ("coverage_pct", "coverage %", true),
        ("estimate_complexity_score", "complexity score", true),
        ("stage_count", "stage count", true),
        ("tokens.cache_read", "cache read", true),
        ("tokens.cache_creation", "cache creation", true),
        ("pass_fail", "pass/fail", false),
    ]

    static func htmlEscape(_ s: String) -> String {
        var r = ""
        for ch in s {
            switch ch {
            case "&": r += "&amp;"
            case "<": r += "&lt;"
            case ">": r += "&gt;"
            case "\"": r += "&quot;"
            case "'": r += "&#x27;"
            default: r.append(ch)
            }
        }
        return r
    }

    static func get(_ pathMetrics: JSONValue?, _ accessor: String) -> JSONValue? {
        guard let pathMetrics else { return nil }
        if accessor.contains(".") {
            let parts = accessor.split(separator: ".", maxSplits: 1).map(String.init)
            return pathMetrics[parts[0]]?[parts[1]]
        }
        return pathMetrics[accessor]
    }

    /// Format a scalar JSONValue for a table cell (Python `_fmt`).
    static func fmt(_ v: JSONValue?) -> String {
        guard let v, !v.isNull else { return "&mdash;" }
        switch v {
        case .bool(let b): return b ? "pass" : "fail"
        case .int(let i): return groupInt(i)
        case .double(let d):
            if d == d.rounded() { return groupInt(Int(d)) }
            // 4dp then strip trailing zeros + dot.
            var s = String(format: "%.4f", d)
            while s.hasSuffix("0") { s.removeLast() }
            if s.hasSuffix(".") { s.removeLast() }
            return groupDecimal(s)
        case .string(let s): return htmlEscape(s)
        default: return htmlEscape(v.serialized(indent: 0))
        }
    }

    static func groupInt(_ i: Int) -> String {
        let neg = i < 0
        var digits = Array(String(abs(i)))
        var out: [Character] = []
        var c = 0
        for d in digits.reversed() {
            if c != 0 && c % 3 == 0 { out.append(",") }
            out.append(d); c += 1
        }
        digits = out.reversed()
        return (neg ? "-" : "") + String(digits)
    }

    static func groupDecimal(_ s: String) -> String {
        let parts = s.split(separator: ".", maxSplits: 1).map(String.init)
        guard let intPart = parts.first, let n = Int(intPart) else { return s }
        let grouped = groupInt(n)
        return parts.count == 2 ? "\(grouped).\(parts[1])" : grouped
    }

    static func delta(_ withV: JSONValue?, _ withoutV: JSONValue?) -> String {
        func num(_ v: JSONValue?) -> Double? {
            switch v {
            case .int(let i): return Double(i)
            case .double(let d): return d
            default: return nil
            }
        }
        if let w = num(withV), let wo = num(withoutV) {
            let d = w - wo
            let sign = d > 0 ? "+" : ""
            let dv: JSONValue = (d == d.rounded()) ? .int(Int(d)) : .double(d)
            return "\(sign)\(fmt(dv))"
        }
        return "&mdash;"
    }

    static func appPath(_ rec: JSONValue, _ which: String) -> String {
        let pm = rec["paths"]?[which]
        if let p = pm?["app_path"]?.stringValue, !p.isEmpty { return p }
        let runID = rec["run_id"]?.stringValue ?? "unknown"
        return "benchmark/workdirs/\(runID)/\(which)"
    }

    static func numOrNil(_ v: JSONValue?) -> Double? {
        switch v {
        case .int(let i): return Double(i)
        case .double(let d): return d
        default: return nil
        }
    }

    static func ratioFmt(_ v: Double) -> String {
        var s = String(format: "%.2f", v)
        while s.hasSuffix("0") { s.removeLast() }
        if s.hasSuffix(".") { s.removeLast() }
        return groupDecimal(s)
    }

    /// Derived analysis rows (Python `_analysis_rows`). Each degrades to em-dash.
    static func analysisRows(_ rec: JSONValue) -> [(String, String)] {
        let withPM = rec["paths"]?["with"]
        let withoutPM = rec["paths"]?["without"]
        let wt = withPM?["tokens"]
        let ot = withoutPM?["tokens"]

        let wTotal = numOrNil(wt?["total"])
        let wOut = numOrNil(wt?["out"])
        let wFresh = numOrNil(wt?["in"])
        let wCR = numOrNil(wt?["cache_read"])
        let wCC = numOrNil(wt?["cache_creation"])
        let wLOC = numOrNil(withPM?["loc_produced"])
        let oTotal = numOrNil(ot?["total"])

        var rows: [(String, String)] = []

        // input:output ratio
        if let wTotal, let wOut, wOut != 0 {
            rows.append(("input:output ratio", "\(ratioFmt(wTotal / wOut)):1"))
        } else { rows.append(("input:output ratio", "&mdash;")) }

        // cache-hit %
        if let wCR, let wCC, let wFresh {
            let denom = wFresh + wCC + wCR
            rows.append(("cache-hit %", denom != 0 ? "\(ratioFmt(wCR / denom * 100))%" : "&mdash;"))
        } else { rows.append(("cache-hit %", "&mdash;")) }

        // per-stage token share
        let stages = rec["stages"]?.arrayItems ?? []
        var stageTotals: [(String, Int)] = []
        for s in stages {
            let fi = numOrNil(s["fresh_in"]).map(Int.init) ?? 0
            let cc = numOrNil(s["cache_creation"]).map(Int.init) ?? 0
            let cr = numOrNil(s["cache_read"]).map(Int.init) ?? 0
            stageTotals.append((s["stage"]?.stringValue ?? "?", fi + cc + cr))
        }
        let grand = stageTotals.reduce(0) { $0 + $1.1 }
        if !stageTotals.isEmpty && grand != 0 {
            let parts = stageTotals.map { "\(htmlEscape($0.0)) \(ratioFmt(Double($0.1) / Double(grand) * 100))%" }
            rows.append(("per-stage token share", parts.joined(separator: " &middot; ")))
        } else { rows.append(("per-stage token share", "&mdash;")) }

        // tokens per LOC
        if let wTotal, let wLOC, wLOC != 0 {
            rows.append(("tokens per LOC", ratioFmt(wTotal / wLOC)))
        } else { rows.append(("tokens per LOC", "&mdash;")) }

        // WITH-vs-WITHOUT premium
        if let wTotal, let oTotal, oTotal != 0 {
            rows.append(("WITH-vs-WITHOUT premium", "\(ratioFmt((wTotal - oTotal) / oTotal * 100))%"))
        } else { rows.append(("WITH-vs-WITHOUT premium", "&mdash;")) }

        return rows
    }

    /// Coverage rollup rows for a live record carrying a per-stage coverage
    /// manifest — distinct agents / skills / commands / total tool_calls across
    /// all stages. Empty when no stage has a coverage object.
    static func coverageRows(_ rec: JSONValue) -> [(String, String)] {
        var agents = Set<String>(), skills = Set<String>(), commands = Set<String>()
        var toolCalls = 0
        var any = false
        for s in rec["stages"]?.arrayItems ?? [] {
            guard let cov = s["coverage"], case .object = cov else { continue }
            any = true
            (cov["agents"]?.arrayItems ?? []).compactMap(\.stringValue).forEach { agents.insert($0) }
            (cov["skills"]?.arrayItems ?? []).compactMap(\.stringValue).forEach { skills.insert($0) }
            (cov["commands"]?.arrayItems ?? []).compactMap(\.stringValue).forEach { commands.insert($0) }
            toolCalls += cov["tool_calls"]?.intValue ?? 0
        }
        guard any else { return [] }
        return [
            ("agents exercised", agents.isEmpty ? "&mdash;" : htmlEscape(agents.sorted().joined(separator: ", "))),
            ("skills exercised", skills.isEmpty ? "&mdash;" : htmlEscape(skills.sorted().joined(separator: ", "))),
            ("commands exercised", commands.isEmpty ? "&mdash;" : htmlEscape(commands.sorted().joined(separator: ", "))),
            ("tool calls", String(toolCalls)),
        ]
    }

    /// Count the plugin's declared surface (agents/commands/skills) for the
    /// exercised/declared ratio. Best-effort; missing dirs -> 0.
    public static func declaredSurface(pluginRoot: String) -> (agents: Int, skills: Int, commands: Int) {
        let fm = FileManager.default
        func countMD(_ dir: String) -> Int {
            ((try? fm.contentsOfDirectory(atPath: dir)) ?? []).filter { $0.hasSuffix(".md") }.count
        }
        let agents = countMD((pluginRoot as NSString).appendingPathComponent("agents"))
        let commands = countMD((pluginRoot as NSString).appendingPathComponent("commands"))
        // skills/*/SKILL.md
        var skills = 0
        let skillsDir = (pluginRoot as NSString).appendingPathComponent("skills")
        for sub in (try? fm.contentsOfDirectory(atPath: skillsDir)) ?? [] {
            let p = (skillsDir as NSString).appendingPathComponent(sub) + "/SKILL.md"
            if fm.fileExists(atPath: p) { skills += 1 }
        }
        return (agents, skills, commands)
    }

    static func recordHTML(_ rec: JSONValue) -> String {
        let runID = htmlEscape(rec["run_id"]?.stringValue ?? "?")
        let ts = htmlEscape(rec["timestamp_utc"]?.stringValue ?? "?")
        let mode = htmlEscape(rec["mode"]?.stringValue ?? "?")
        let sha = htmlEscape(rec["git_sha"]?.stringValue ?? "?")
        let budget = rec["budget_usd"]
        let partial = rec["live_partial"]?.boolValue ?? false
        let withPM = rec["paths"]?["with"]
        let withoutPM = rec["paths"]?["without"]

        var rows = ""
        for (accessor, label, _) in metricRows {
            let wv = get(withPM, accessor)
            let ov = get(withoutPM, accessor)
            rows += "<tr><td class='metric'>\(htmlEscape(label))</td>"
                + "<td class='with'>\(fmt(wv))</td>"
                + "<td class='without'>\(fmt(ov))</td>"
                + "<td class='delta'>\(delta(wv, ov))</td></tr>"
        }

        var badges = "<span class='badge mode-\(mode)'>\(mode)</span>"
        if let budget, !budget.isNull { badges += "<span class='badge budget'>budget $\(fmt(budget))</span>" }
        if partial { badges += "<span class='badge partial'>live_partial</span>" }

        let withApp = htmlEscape(appPath(rec, "with"))
        let withoutApp = htmlEscape(appPath(rec, "without"))

        let analysis = analysisRows(rec).map {
            "<tr><td class='metric'>\(htmlEscape($0.0))</td><td class='analysis' colspan='3'>\($0.1)</td></tr>"
        }.joined()

        let coverage = coverageRows(rec)
        var coverageSection = ""
        if !coverage.isEmpty {
            let covBody = coverage.map {
                "<tr><td class='metric'>\(htmlEscape($0.0))</td><td class='analysis' colspan='3'>\($0.1)</td></tr>"
            }.joined()
            coverageSection = """
              <table class="analysis-table">
                <thead><tr><th>coverage manifest</th><th colspan="3">exercised</th></tr></thead>
                <tbody>\(covBody)</tbody>
              </table>
            """
        }

        return """
            <section class="run">
              <h3>\(runID) \(badges)</h3>
              <div class="meta">\(ts) &middot; git \(sha)</div>
              <table>
                <thead><tr><th>metric</th><th>WITH plugin</th><th>WITHOUT plugin</th><th>&Delta; (with&minus;without)</th></tr></thead>
                <tbody>\(rows)</tbody>
              </table>
              <table class="analysis-table">
                <thead><tr><th>analysis (derived)</th><th colspan="3">value</th></tr></thead>
                <tbody>\(analysis)</tbody>
              </table>
              \(coverageSection)
              <div class="apps">
                <div><span class="lbl">WITH app:</span> <code>\(withApp)</code></div>
                <div><span class="lbl">WITHOUT app:</span> <code>\(withoutApp)</code></div>
              </div>
            </section>
        """
    }

    static let css = """
    :root { --with:#2563eb; --without:#64748b; --pos:#16a34a; --bg:#0b1020; --card:#141b2e; --fg:#e5e9f0; }
    * { box-sizing:border-box; } body { margin:0; font:14px/1.5 -apple-system,Segoe UI,Roboto,sans-serif;
      background:var(--bg); color:var(--fg); padding:24px; }
    h1 { font-size:20px; margin:0 0 4px; } h2 { font-size:16px; margin:28px 0 8px; border-bottom:1px solid #26304a; padding-bottom:4px; }
    .sub { color:#94a3b8; margin:0 0 20px; }
    .run { background:var(--card); border:1px solid #26304a; border-radius:10px; padding:16px; margin:12px 0; }
    .run h3 { margin:0 0 2px; font-size:14px; font-family:ui-monospace,Menlo,monospace; }
    .meta { color:#94a3b8; font-size:12px; margin-bottom:10px; }
    table { width:100%; border-collapse:collapse; }
    th,td { text-align:right; padding:5px 10px; border-bottom:1px solid #232c44; }
    th:first-child, td.metric { text-align:left; color:#cbd5e1; }
    td.with { color:#93b4fb; } td.without { color:#cbd5e1; } td.delta { color:#86efac; font-variant-numeric:tabular-nums; }
    td.analysis { color:#fcd34d; text-align:right; font-variant-numeric:tabular-nums; }
    .analysis-table { margin-top:8px; } .analysis-table th { color:#fbbf24; }
    td { font-variant-numeric:tabular-nums; }
    .apps { margin-top:12px; font-size:13px; } .apps .lbl { color:#94a3b8; display:inline-block; width:90px; }
    .apps code { background:#0b1224; padding:2px 6px; border-radius:5px; color:#e2e8f0; }
    .badge { font-size:11px; padding:2px 7px; border-radius:999px; margin-left:8px; vertical-align:middle; }
    .mode-deterministic { background:#1e3a5f; color:#93c5fd; } .mode-live { background:#4a1d3f; color:#f0abfc; }
    .budget { background:#3a2f0b; color:#fde047; } .partial { background:#4a1d1d; color:#fca5a5; }
    .empty { color:#94a3b8; }
    footer { color:#64748b; margin-top:28px; font-size:12px; }
    """

    public static func renderHTML(history: JSONValue, pluginRoot: String? = nil) -> String {
        var bodyParts: [String] = []
        var latestTS = ""
        for mode in ["deterministic", "live"] {
            guard let recs = history[mode]?.arrayItems, !recs.isEmpty else { continue }
            bodyParts.append("<h2>\(htmlEscape(mode)) &mdash; latest \(recs.count)</h2>")
            for rec in recs.reversed() {
                if latestTS.isEmpty { latestTS = rec["timestamp_utc"]?.stringValue ?? "" }
                bodyParts.append(recordHTML(rec))
            }
        }
        if bodyParts.isEmpty {
            bodyParts.append("<p class='empty'>No benchmark results yet. Run <code>make benchmark</code>.</p>")
        }

        // Plugin-surface coverage ratio (exercised/declared), computed at report
        // time. exercised = distinct agents/skills/commands across all live
        // records' coverage manifests.
        var surfaceLine = ""
        if let pluginRoot {
            let declared = declaredSurface(pluginRoot: pluginRoot)
            var exAgents = Set<String>(), exSkills = Set<String>(), exCommands = Set<String>()
            for rec in (history["live"]?.arrayItems ?? []) {
                for s in rec["stages"]?.arrayItems ?? [] {
                    guard let cov = s["coverage"], case .object = cov else { continue }
                    (cov["agents"]?.arrayItems ?? []).compactMap(\.stringValue).forEach { exAgents.insert($0) }
                    (cov["skills"]?.arrayItems ?? []).compactMap(\.stringValue).forEach { exSkills.insert($0) }
                    (cov["commands"]?.arrayItems ?? []).compactMap(\.stringValue).forEach { exCommands.insert($0) }
                }
            }
            func ratio(_ ex: Int, _ decl: Int) -> String {
                decl > 0 ? "\(ex)/\(decl) (\(ratioFmt(Double(ex) / Double(decl) * 100))%)" : "\(ex)/0"
            }
            surfaceLine = "<p class='sub'>Plugin-surface coverage — agents "
                + "\(ratio(exAgents.count, declared.agents)) &middot; skills "
                + "\(ratio(exSkills.count, declared.skills)) &middot; commands "
                + "\(ratio(exCommands.count, declared.commands)).</p>"
        }

        let asof = latestTS.isEmpty ? "no runs" : htmlEscape(latestTS)
        return """
        <!DOCTYPE html>
        <html lang="en"><head><meta charset="utf-8">
        <meta name="viewport" content="width=device-width, initial-scale=1">
        <title>igrsoft benchmark &mdash; Tic-Tac-Toe with/without plugin</title>
        <style>\(css)</style></head>
        <body>
          <h1>Tic-Tac-Toe benchmark &mdash; WITH vs WITHOUT plugin</h1>
          <p class="sub">All metrics per retained run (latest-3 per mode). As of \(asof). Deterministic tokens/cost are null by design; live fills them.</p>
          \(surfaceLine)
          \(bodyParts.joined())
          <footer>Generated by bench-report from benchmark/results/history.json &middot; app paths point to gitignored per-run workdirs.</footer>
        </body></html>

        """
    }

    /// Read history at `historyPath`, render, and write `outPath`. Returns outPath.
    @discardableResult
    public static func buildReport(historyPath: String, outPath: String,
                                   pluginRoot: String? = nil) throws -> String {
        var history: JSONValue = .object([])
        if let text = try? String(contentsOfFile: historyPath, encoding: .utf8),
           let parsed = try? JSONParser.parse(text), case .object = parsed {
            history = parsed
        }
        let htmlText = renderHTML(history: history, pluginRoot: pluginRoot)
        let outDir = (outPath as NSString).deletingLastPathComponent
        try FileManager.default.createDirectory(atPath: outDir, withIntermediateDirectories: true)
        try htmlText.write(toFile: outPath, atomically: true, encoding: .utf8)
        return outPath
    }
}

File: benchmark/harness/benchmarkkit/report.py
"""Render benchmark/results/history.json into a self-contained result.html.

Shows all metrics (WITH vs WITHOUT + deltas) per retained run/mode, newest first,
each app path, derived analysis rows (each degrades to an em-dash on missing
inputs), coverage rollup rows for live records, and a plugin-surface coverage
ratio. Operates on the raw parsed history dict. HTML is substring-gated, not
byte-gated.
"""

from __future__ import annotations

import html as _html
import os
from typing import Optional

from .analysis import scan_generated_project

# (accessor, label, is_delta_numeric)
METRIC_ROWS = [
    ("tokens.in", "tokens in", True),
    ("tokens.out", "tokens out", True),
    ("tokens.total", "tokens total", True),
    ("cost_usd", "cost (USD)", True),
    ("wall_clock_s", "wall clock (s)", True),
    ("loc_produced", "LOC produced", True),
    ("test_count", "test count", True),
    ("coverage_pct", "coverage %", True),
    ("estimate_complexity_score", "complexity score", True),
    ("stage_count", "stage count", True),
    ("tokens.cache_read", "cache read", True),
    ("tokens.cache_creation", "cache creation", True),
    ("pass_fail", "pass/fail", False),
]


def _escape(s: str) -> str:
    return _html.escape(s, quote=True)


def _get(path_metrics: Optional[dict], accessor: str):
    if path_metrics is None:
        return None
    if "." in accessor:
        a, b = accessor.split(".", 1)
        inner = path_metrics.get(a)
        return inner.get(b) if isinstance(inner, dict) else None
    return path_metrics.get(accessor)


def _group_int(i: int) -> str:
    neg = i < 0
    return ("-" if neg else "") + f"{abs(i):,}"


def _group_decimal(s: str) -> str:
    parts = s.split(".", 1)
    try:
        n = int(parts[0])
    except (ValueError, IndexError):
        return s
    grouped = _group_int(n)
    return f"{grouped}.{parts[1]}" if len(parts) == 2 else grouped


def _fmt(v) -> str:
    if v is None:
        return "&mdash;"
    if isinstance(v, bool):
        return "pass" if v else "fail"
    if isinstance(v, int):
        return _group_int(v)
    if isinstance(v, float):
        if v == round(v):
            return _group_int(int(v))
        s = f"{v:.4f}".rstrip("0").rstrip(".")
        return _group_decimal(s)
    return _escape(str(v))


def _num_or_nil(v):
    if isinstance(v, bool):
        return None
    if isinstance(v, (int, float)):
        return float(v)
    return None


def _delta(with_v, without_v) -> str:
    w = _num_or_nil(with_v)
    wo = _num_or_nil(without_v)
    if w is not None and wo is not None:
        d = w - wo
        sign = "+" if d > 0 else ""
        dv = int(d) if d == round(d) else d
        return f"{sign}{_fmt(dv)}"
    return "&mdash;"


def _app_path(rec: dict, which: str) -> str:
    pm = (rec.get("paths") or {}).get(which) or {}
    p = pm.get("app_path")
    if p:
        return p
    run_id = rec.get("run_id") or "unknown"
    return f"benchmark/workdirs/{run_id}/{which}"


def _ratio_fmt(v: float) -> str:
    s = f"{v:.2f}".rstrip("0").rstrip(".")
    return _group_decimal(s)


def analysis_rows(rec: dict) -> list:
    with_pm = (rec.get("paths") or {}).get("with") or {}
    without_pm = (rec.get("paths") or {}).get("without") or {}
    wt = with_pm.get("tokens") or {}
    ot = without_pm.get("tokens") or {}

    w_total = _num_or_nil(wt.get("total"))
    w_out = _num_or_nil(wt.get("out"))
    w_fresh = _num_or_nil(wt.get("in"))
    w_cr = _num_or_nil(wt.get("cache_read"))
    w_cc = _num_or_nil(wt.get("cache_creation"))
    w_loc = _num_or_nil(with_pm.get("loc_produced"))
    o_total = _num_or_nil(ot.get("total"))

    rows = []

    if w_total is not None and w_out not in (None, 0):
        rows.append(("input:output ratio", f"{_ratio_fmt(w_total / w_out)}:1"))
    else:
        rows.append(("input:output ratio", "&mdash;"))

    if w_cr is not None and w_cc is not None and w_fresh is not None:
        denom = w_fresh + w_cc + w_cr
        rows.append(("cache-hit %", f"{_ratio_fmt(w_cr / denom * 100)}%" if denom != 0 else "&mdash;"))
    else:
        rows.append(("cache-hit %", "&mdash;"))

    stage_totals = []
    for s in rec.get("stages") or []:
        fi = int(_num_or_nil(s.get("fresh_in")) or 0)
        cc = int(_num_or_nil(s.get("cache_creation")) or 0)
        cr = int(_num_or_nil(s.get("cache_read")) or 0)
        stage_totals.append((s.get("stage") or "?", s.get("arm"), fi + cc + cr))
    grand = sum(t for _, _, t in stage_totals)
    if stage_totals and grand != 0:
        # Arm tag disambiguates the paired path, where each stage name appears once per arm.
        parts = []
        for name, arm, t in stage_totals:
            label = f"{name} ({arm})" if arm in ("with", "without") else name
            parts.append(f"{_escape(label)} {_ratio_fmt(t / grand * 100)}%")
        rows.append(("per-stage token share", " &middot; ".join(parts)))
    else:
        rows.append(("per-stage token share", "&mdash;"))

    if w_total is not None and w_loc not in (None, 0):
        rows.append(("tokens per LOC", _ratio_fmt(w_total / w_loc)))
    else:
        rows.append(("tokens per LOC", "&mdash;"))

    if w_total is not None and o_total not in (None, 0):
        rows.append(("WITH-vs-WITHOUT premium", f"{_ratio_fmt((w_total - o_total) / o_total * 100)}%"))
    else:
        rows.append(("WITH-vs-WITHOUT premium", "&mdash;"))

    return rows


def coverage_rows(rec: dict) -> list:
    agents, skills, commands = set(), set(), set()
    tool_calls = 0
    any_cov = False
    for s in rec.get("stages") or []:
        cov = s.get("coverage")
        if not isinstance(cov, dict):
            continue
        any_cov = True
        agents.update(x for x in cov.get("agents", []) if isinstance(x, str))
        skills.update(x for x in cov.get("skills", []) if isinstance(x, str))
        commands.update(x for x in cov.get("commands", []) if isinstance(x, str))
        tool_calls += cov.get("tool_calls") or 0
    nested = 0
    for s in rec.get("stages") or []:
        cov = s.get("coverage")
        if isinstance(cov, dict):
            nested += cov.get("nested_background") or 0
    if not any_cov:
        return []
    rows = [
        ("agents exercised", _escape(", ".join(sorted(agents))) if agents else "&mdash;"),
        ("skills exercised", _escape(", ".join(sorted(skills))) if skills else "&mdash;"),
        ("commands exercised", _escape(", ".join(sorted(commands))) if commands else "&mdash;"),
        ("tool calls", str(tool_calls)),
    ]
    if nested:
        rows.append(("nested background spawns", str(nested)))
    return rows


def _generated_html(rec: dict, plugin_root: Optional[str]) -> str:
    """Per-arm Generated-project section (U2): Swift file tree, per-file LOC, total,
    arm path. Resolved from each arm's app_path under plugin_root; empty when absent."""
    if not plugin_root:
        return ""
    blocks = []
    for arm in ("with", "without"):
        pm = (rec.get("paths") or {}).get(arm) or {}
        app_path = pm.get("app_path")
        if not app_path:
            continue
        scan = scan_generated_project(os.path.join(plugin_root, app_path))
        if scan is None:
            continue
        rows = "".join(
            f"<tr><td class='metric'><code>{_escape(rel)}</code></td>"
            f"<td class='analysis' colspan='3'>{loc}</td></tr>"
            for rel, loc in scan["files"]
        )
        blocks.append(
            '<table class="analysis-table"><thead><tr>'
            f"<th>generated project ({_escape(arm)}) &mdash; "
            f"<code>{_escape(app_path)}</code>, total LOC {scan['total_loc']}</th>"
            '<th colspan="3">LOC</th></tr></thead>'
            f"<tbody>{rows}</tbody></table>"
        )
    return "".join(blocks)


def declared_surface(plugin_root: str) -> tuple:
    def count_md(d: str) -> int:
        try:
            return len([n for n in os.listdir(d) if n.endswith(".md")])
        except OSError:
            return 0

    agents = count_md(os.path.join(plugin_root, "agents"))
    commands = count_md(os.path.join(plugin_root, "commands"))
    skills = 0
    skills_dir = os.path.join(plugin_root, "skills")
    try:
        for sub in os.listdir(skills_dir):
            if os.path.exists(os.path.join(skills_dir, sub, "SKILL.md")):
                skills += 1
    except OSError:
        pass
    return agents, skills, commands


def _record_html(rec: dict, plugin_root: Optional[str] = None) -> str:
    run_id = _escape(rec.get("run_id") or "?")
    ts = _escape(rec.get("timestamp_utc") or "?")
    mode = _escape(rec.get("mode") or "?")
    sha = _escape(rec.get("git_sha") or "?")
    budget = rec.get("budget_usd")
    partial = bool(rec.get("live_partial") or False)
    with_pm = (rec.get("paths") or {}).get("with")
    without_pm = (rec.get("paths") or {}).get("without")

    rows = ""
    for accessor, label, _ in METRIC_ROWS:
        wv = _get(with_pm, accessor)
        ov = _get(without_pm, accessor)
        # Coverage is unmeasured on the live path (both arms None): omit the row so absent
        # never renders as a real 0. A measured 0.0 (deterministic) still shows a 0 row.
        if accessor == "coverage_pct" and wv is None and ov is None:
            continue
        rows += (
            f"<tr><td class='metric'>{_escape(label)}</td>"
            f"<td class='with'>{_fmt(wv)}</td>"
            f"<td class='without'>{_fmt(ov)}</td>"
            f"<td class='delta'>{_delta(wv, ov)}</td></tr>"
        )

    badges = f"<span class='badge mode-{mode}'>{mode}</span>"
    if budget is not None:
        badges += f"<span class='badge budget'>budget ${_fmt(budget)}</span>"
    if partial:
        badges += "<span class='badge partial'>live_partial</span>"

    with_app = _escape(_app_path(rec, "with"))
    without_app = _escape(_app_path(rec, "without"))

    analysis = "".join(
        f"<tr><td class='metric'>{_escape(k)}</td><td class='analysis' colspan='3'>{v}</td></tr>"
        for k, v in analysis_rows(rec)
    )

    coverage = coverage_rows(rec)
    coverage_section = ""
    if coverage:
        cov_body = "".join(
            f"<tr><td class='metric'>{_escape(k)}</td><td class='analysis' colspan='3'>{v}</td></tr>"
            for k, v in coverage
        )
        coverage_section = (
            '<table class="analysis-table">'
            '<thead><tr><th>coverage manifest</th><th colspan="3">exercised</th></tr></thead>'
            f"<tbody>{cov_body}</tbody></table>"
        )

    return (
        '<section class="run">'
        f"<h3>{run_id} {badges}</h3>"
        f'<div class="meta">{ts} &middot; git {sha}</div>'
        "<table><thead><tr><th>metric</th><th>WITH plugin</th><th>WITHOUT plugin</th>"
        "<th>&Delta; (with&minus;without)</th></tr></thead>"
        f"<tbody>{rows}</tbody></table>"
        '<table class="analysis-table"><thead><tr><th>analysis (derived)</th>'
        f'<th colspan="3">value</th></tr></thead><tbody>{analysis}</tbody></table>'
        f"{coverage_section}"
        f"{_generated_html(rec, plugin_root)}"
        '<div class="apps">'
        f'<div><span class="lbl">WITH app:</span> <code>{with_app}</code></div>'
        f'<div><span class="lbl">WITHOUT app:</span> <code>{without_app}</code></div>'
        "</div></section>"
    )


_CSS = (
    ":root{--with:#2563eb;--without:#64748b;--pos:#16a34a;--bg:#0b1020;--card:#141b2e;--fg:#e5e9f0;}"
    "body{margin:0;font:14px/1.5 -apple-system,Segoe UI,Roboto,sans-serif;background:var(--bg);"
    "color:var(--fg);padding:24px;}"
    "table{width:100%;border-collapse:collapse;}"
    "td.delta{color:#86efac;} td.analysis{color:#fcd34d;}"
)


def render_html(history: dict, plugin_root: Optional[str] = None) -> str:
    body_parts = []
    latest_ts = ""
    for mode in ("deterministic", "live"):
        recs = history.get(mode) if isinstance(history, dict) else None
        if not recs:
            continue
        body_parts.append(f"<h2>{_escape(mode)} &mdash; latest {len(recs)}</h2>")
        for rec in reversed(recs):
            if not latest_ts:
                latest_ts = rec.get("timestamp_utc") or ""
            body_parts.append(_record_html(rec, plugin_root=plugin_root))
    if not body_parts:
        body_parts.append("<p class='empty'>No benchmark results yet. Run <code>make benchmark</code>.</p>")

    surface_line = ""
    if plugin_root:
        da, ds, dc = declared_surface(plugin_root)
        ex_agents, ex_skills, ex_commands = set(), set(), set()
        for rec in (history.get("live") or []) if isinstance(history, dict) else []:
            for s in rec.get("stages") or []:
                cov = s.get("coverage")
                if not isinstance(cov, dict):
                    continue
                ex_agents.update(x for x in cov.get("agents", []) if isinstance(x, str))
                ex_skills.update(x for x in cov.get("skills", []) if isinstance(x, str))
                ex_commands.update(x for x in cov.get("commands", []) if isinstance(x, str))

        def ratio(ex: int, decl: int) -> str:
            return f"{ex}/{decl} ({_ratio_fmt(ex / decl * 100)}%)" if decl > 0 else f"{ex}/0"

        surface_line = (
            "<p class='sub'>Plugin-surface coverage — agents "
            f"{ratio(len(ex_agents), da)} &middot; skills "
            f"{ratio(len(ex_skills), ds)} &middot; commands "
            f"{ratio(len(ex_commands), dc)}.</p>"
        )

    asof = _escape(latest_ts) if latest_ts else "no runs"
    return (
        "<!DOCTYPE html>\n"
        '<html lang="en"><head><meta charset="utf-8">'
        '<meta name="viewport" content="width=device-width, initial-scale=1">'
        "<title>corpflow benchmark &mdash; Tic-Tac-Toe with/without plugin</title>"
        f"<style>{_CSS}</style></head><body>"
        "<h1>Tic-Tac-Toe benchmark &mdash; WITH vs WITHOUT plugin</h1>"
        '<p class="sub">All metrics per retained run (latest-3 per mode). As of '
        f"{asof}. Deterministic tokens/cost are null by design; live fills them.</p>"
        f"{surface_line}"
        f"{''.join(body_parts)}"
        "<footer>Generated by bench-report from benchmark/results/history.json &middot; "
        "app paths point to gitignored per-run workdirs.</footer>"
        "</body></html>\n"
    )


def build_report(history_path: str, out_path: str, plugin_root: Optional[str] = None) -> str:
    history = {}
    try:
        import json
        with open(history_path, encoding="utf-8") as f:
            parsed = json.load(f)
        if isinstance(parsed, dict):
            history = parsed
    except (OSError, ValueError):
        history = {}
    html_text = render_html(history, plugin_root=plugin_root)
    out_dir = os.path.dirname(out_path)
    if out_dir:
        os.makedirs(out_dir, exist_ok=True)
    with open(out_path, "w", encoding="utf-8") as f:
        f.write(html_text)
    return out_path
